### ISSUE TITLE
Nested parallel_scan

### ISSUE THREAD

**Issue by dholladay00** [kokkos/kokkos#433]
I would like to use a `parallel_scan` over a `ThreadVectorRange` within a `parallel_scan` over a `TeamThreadRange`. My current understanding is that this would require the innermost scan to be implemented with a lambda and that scans with a lambda are not supported on the Cuda backend. 

Is there a way around this and/or support on the horizon for scans via a lambda on the Cuda backend?

**Comment by crtrott** [comment]
The biggest problem is the scan over the TeamThreadRange. That is not implemented at all (at any backend). The scan at the TeamVectorRange wouldn't be too big of a deal though.

**Comment by dholladay00** [comment]
It can be implemented as:
```c++
Kokkos::single (PerTeam(...), [&] ()
{
  // nested sequential for loops
});
```

Is this something you plan on implementing?
If not, can it be added as a feature request? 
The above will work as it should never dominate the calculation time, but it would be nice to implement in parallel.

**Comment by crtrott** [comment]
Yeah we can look into it. And in fact there is a feature request for it: #45 
As you can see it has been in the backlog forever, mostly because nobody with a real need asked.

**Comment by dholladay00** [comment]
There are a few places where I need scans to calculate offsets for team_scratch memory allocation, what is the timeline on the enhancement?

**Comment by dholladay00** [comment]
I've heard this has been added in the OpenMP backend. Is that worth adding an 'in develop' tag yet?

**Comment by dholladay00** [comment]
I have `TeamThreadRange` parallel scans in which I do:

```c++
Kokkos::parallel_scan (TeamThreadRange(team), [&] (const int i, int& update, const bool last)
{
  Kokkos::single(PerTeam(team), [&] (),
  {
    // do stuff
  });
});
```

Once TeamVectorRange is working, the single could be removed. 

I also have a parallel scan over a thread vector range inside of a parallel for over a team thread range. All of these appear to work correctly on OpenMP and Cuda backends with the develop branch. 

For my purposes at least, this feature is `In Develop`.

**Comment by dholladay00** [comment]
This in fact fails on the Cuda backend: 
```
warning: calling a __host__ function from a __host__ __device__ function is not allowed

/.../kokkos/core/src/Kokkos_Parallel.hpp(429): error: incomplete type is not allowed
          detected during instantiation of "void Kokkos::parallel_scan(const ExecutionPolicy &, const FunctorType &, const std::__cxx11::string &, Kokkos::Impl::enable_if<<expression>, void>::type *) [with ExecutionPolicy=Kokkos::Impl::TeamThreadRangeBoundariesStruct<unsigned long, member_type>, FunctorType=lambda [](int_t, int_t &, __nv_bool)->void]" 
```

So I guess it needs to remain open.

The first warning looks like it might need to be implemented with a functor regardless.

If that's the case, for the functor, would the parentheses operator look like
```c++
operator () (const int i, int& update, const bool last)
```
or
```c++
operator () (const member_type& team, int& update, const bool last)
```
?

**Comment by ibaned** [comment]
Closing as a duplicate of #45, will continue this effort there.
